lm_lstm_ptb.py

```diff
@@ -0,0 +1,900 @@
+def adam(lr, tparams, gshared, beta1=0.9, beta2=0.99, e=1e-5):
+    updates = []
+    t_prev = theano.shared(numpy.float32(0.))
+    t = t_prev + 1.
+    lr_t = lr * tensor.sqrt(1. - beta2**t) / (1. - beta1**t)
+    for p, g in zip(tparams.values(), gshared):
+        m = theano.shared(p.get_value() * 0., p.name + '_mean')
+        v = theano.shared(p.get_value() * 0., p.name + '_variance')
+        m_t = beta1 * m + (1. - beta1) * g
+        v_t = beta2 * v + (1. - beta2) * g**2
+        step = lr_t * m_t / (tensor.sqrt(v_t) + e)
+        p_t = p - step
+        updates.append((m, m_t))
+        updates.append((v, v_t))
+        updates.append((p, p_t))
+    updates.append((t_prev, t))
+    f_update = theano.function([lr], [], updates=updates, profile=profile)
+    return f_update
+
+
+def train(dim_word=200,  # input vector dimensionality
+          dim=2000,  # the number of GRU units
+          dim_proj=600,  # the number of GRU units
+          encoder='lstm',
+          patience=10,  # early stopping patience
+          max_epochs=5000,
+          finish_after=10000000,  # finish after this many updates
+          dispFreq=100,
+          decay_c=0.,  # L2 weight decay penalty
+          lrate=0.001,
+          maxlen=100,  # maximum length of the description
+          optimizer='adam',
+          batch_size=16,
+          n_words=30000,
+          valid_batch_size=16,
+          saveto='model.npz',
+          validFreq=1000,
+          saveFreq=1000,  # save the parameters after every saveFreq updates
+          sampleFreq=100,  # generate some samples after every sampleFreq
+          dataset=None,  # Not used
+          valid_dataset=None,  # Not used
+          dictionary=None,  # Not used
+          use_dropout=False,
+          reload_=False,
+          kl_start=0.2,
+          weight_aux=0.,
+          kl_rate=0.0003):
+    dim_proj = dim
+    data_path = '/data/lisatmp4/anirudhg/ptb/'
+    prior_hidden = dim
+    dim_z = 256
+    encoder_hidden = dim
+    learn_h0 = False
+
+    desc = saveto + 'seed_' + str(seed) + '_model_' + str(weight_aux) + '_weight_aux_' +  str(kl_start) + '_kl_Start_' + str(kl_rate) +  '_kl_rate_log.txt'
+    opts = saveto + 'seed_' + str(seed) + '_model_' + str(weight_aux) + '_weight_aux_' +  str(kl_start) + '_kl_Start_' + str(kl_rate) +  '_kl_rate_opts.pkl'
+
+    print(desc)
+
+    raw_data = reader.ptb_raw_data(data_path)
+    train_data, valid_data, test_data, _ = raw_data
+
+
+    # Model options
+    model_options = locals().copy()
+    pkl.dump(model_options, open(opts, 'wb'))
+    log_file = open(desc, 'w')
+
+    #data = TimitData("timit_raw_batchsize64_seqlen40.npz", batch_size=model_options['batch_size'])
+
+    print('Building model')
+    params = init_params(model_options)
+    tparams = init_tparams(params)
+
+    x = tensor.matrix('x', dtype='int64')
+    x_mask = tensor.matrix('x_mask', dtype='float32')
+    y = tensor.matrix('y', dtype='int64')
```

Comment:
Do you need a variable `y` ? I don't remember seeing one in build_model_rev nor build_model_gen.